### ISSUE TITLE
DTLS enhanced cluster support

### ISSUE THREAD

**Issue by kiemlicz** [eclipse/californium#2031]
Hello,

In my company we're interested in connection sharing mechanism
and it _looks like_ we will have some time to invest

Some past issue to bring back some discussions: https://github.com/eclipse/californium/issues/1484

#### Reasoning
The assumption
```
I'm not sure, how many vm/node crashes you have been affected. In my experience, these are very rare, once or twice a year per system
```
is simply wrong. The DTLS state loss happens more often and this is a problem for us.
The graceful shutdown is a nice feature (thank you) however we need more reliability

#### (Very) Roughly we would like to introduce:
1. Storage to keep the CID to NODE_ID mapping (which node handles given CID), populated on client first contact
2. The in-memory state cache, periodically (not with every DTLS packet) flushed to persistent store (only the state change is flushed)
3. Given the node failure occurs (the packet is handled by the node not listed as 'designated' by given CID-NODE_ID mapping),
transfer the state to new node, update CID-NODE_IP mapping

Before starting the work, we would like to have some information regarding:
- are there any particular requirements that we should fulfill in order to get such PR merged into main branch?
- is the JDK7 requirement still valid?

Kind regards

**Comment by boaks** [comment]
> In my experience, these are very rare, once or twice a year per system

> is simply wrong. The DTLS state loss happens more often and this is a problem for us.

What is wrong? My experience?
Obviously not !!!! I run about 4 servers on my own, I had not downtime in the last 1,5 years. There was one move of the californium sandbox to a new vm, but except of that, no I had no downtimes. 

So, how often does your VM crash?
Or are there any other reasons for lose the DTLS state?

> Storage to keep the CID to NODE_ID mapping (which node handles given CID), populated on client first contact

So, you don't want to encode the NODE_ID in the CID? OK then use and maintain a map. The `NodeConnectionIdGenerator`should enable you to do so.

> The in-memory state cache, periodically (not with every DTLS packet) flushed to persistent store (only the state change is flushed)

"(not every DTLS packet)", so you deal with the record sequence number on your own? OK, that record sequence number (write and read) is the only data changing after the handshake. You get callbacks for these changes in `ConnectionListener.onConnectionUpdatesSequenceNumbers(Connection connection, boolean writeSequenceNumber)`, that should do it. Just mark changes (read/write) and then flush these changes, when ever you think it's the right time.

> is the JDK7 requirement still valid?

Until now, JDK7 was working, and so we haven't been forced to update that.
 
> are there any particular requirements that we should fulfill in order to get such PR merged into main branch?

Not sure, if I get that question.

As long as your code is an "alternative", and no critical changes are required, I'm/we're happy to merge it.
Or what your assumptions about that merge?

If it removes other, already working functions, then it's more discussion with all interested ones. And it would cause much more focus on  the quality of the PR. As far as I can see, that should work with the APIs, (or at least with small changes). If that approach works in reality at all, that's your job to make it working.

**Comment by boaks** [comment]
Any news?
Or can I close it?

**Comment by boaks** [comment]
@kiemlicz 

Don't hesitate to comment on this issue or create a new issue, if you want to continue with this.
